src/matching/Generic_vs_generic.ml

```diff
@@ -1075,10 +1075,43 @@ and m_expr ?(is_root = false) a b =
 (* Require an exact match between raw trees except for Any nodes which
    are matched normally. *)
 and m_raw_tree (a : G.raw_tree) (b : G.raw_tree) =
-  match (a, b) with
+  Trace_matching.(if on then print_raw_pair a b);
+  match (a, b) with
+  (* metavar:
+   * We add those extra (Case (_)) or-pattern below because the metavar
+   * could be nested but we want to match at the toplevel.
+   * For instance, Case("Sym_lit", Token("$X")) should also match
+   * Case("Num_lit", Token(("2"))).
+   * TODO? When should we stop? use >||> and try at each level?
+   *)
+  | ( ( Token (str, tok)
+      | Case (_, Token (str, tok))
+      | Case (_, Case (_, Token (str, tok))) ),
+      b )
+    when MV.is_metavar_name str ->
+      envf (str, tok) (MV.Raw b)
+  (* dots: on string.
+   * TODO: we should use m_string_ellipsis_or_metavar_or_default,
+   * but this would require to remove the enclosing quotes first
+   *)
+  | Token ("\"...\"", _), Token (sb, _) when sb =~ "^\".*\"$" -> return ()
   | Token a, Token b -> m_wrap m_string a b
-  | List a, List b -> m_list m_raw_tree a b
-  | Tuple a, Tuple b -> m_list m_raw_tree a b
+  (* dots:
+   * TODO? restrict it to just List?
+   * TODO? we should also accept Any (Expr (Ellipsis)) below. Indeed,
+   * once you start to migrate out of the raw_tree to the real generic AST,
+   * you could get a mix of Raw_tree and generic AST that we need to match
+   * together.
+   *)
+  | List a, List b
+  | Tuple a, Tuple b ->
+      m_list_with_dots m_raw_tree
+        (function
+          | Token ("...", _) -> true
+          | Case (_, Token ("...", _)) -> true
+          | Case (_, Case (_, Token ("...", _))) -> true
+          | _ -> false)
+        ~less_is_ok:false (* empty list can not match non-empty list *) a b
```

Comment:
1. I'm worried that raw-tree lists don't correspond to lists in the original syntax. I think it's gonna be a problem for repeat1(x) constructs that compile to seq(x, repeat(x)) (= Tuple [x; List x]). Maybe that's the only problem, in which case we could do some work on the ocaml-tree-sitter side to cover any seq(x, repeat(x)) into repeat(x). This would lose the static guarantee that the list has at least one element, but it's usually not something we care about anyway.
2. I don't see how tuples matching ellipses is going to work but we'll see. If that works, maybe we don't have to worry about point (1).